### DISCUSSION TITLE
How to design how to show an Entity

### DISCUSSION THREAD

**Discussion by RileyGe** [Macad3D/Macad3D#27] [Q&A] [answered]
Me again. 
I want to add something like a tag to show the position of a vertex. It is something like the image below but the content is the position of a point.

![image](https://user-images.githubusercontent.com/26666371/192149555-a516aedd-c44a-4a18-bb07-a6f9f56f9c4e.png)

I think this function is something like the datum plane. But I can find the code display the datum plane, where I can find them?

**Comment by dipts** [comment, marked as answer]
The managed code is in [VisualPlane.cs](https://github.com/Macad3D/Macad3D/blob/main/Source/Macad.Interaction/Visual/VisualPlane.cs), the native AIS code in [AISX_Plane.cpp](https://github.com/Macad3D/Macad3D/blob/main/Source/Macad.Managed/AisExtensions/AISX_Plane.cpp).

**Reply by RileyGe** [reply]
It's long time since my question. Because I am not familiar with OCC's AIS packages, I looked into these packages before I dig deepper into this question. There are two more concerns about this question:
1. I found a class `AIS_Plane` in OCC `TKV3d` package, why we do not use this class?
2. I am not familiar with C++, can I use C# to implement the class `AISX_Plane`. Is there any goods or bads when using C#?

Thanks

**Reply by dipts** [reply]
1. I needed functionality which is not provided by AIS_Plane. 
2. You need a native class inherited from AIS_InteractiveObject, so C++ is the only reasonable option here.

**Reply by RileyGe** [reply]
Thank you very much.

In my understanding, if I want to implement a new **entity** like DatumPlane I should do three things:

1. An `InteractiveEntity` class. I implement a very simple class in [here](https://github.com/RileyGe/Macad3D/blob/main/Source/Macad.Core/Auxiliary/TextLabel.cs), this class only implement interface `ITransformable` all the same as `DatumPlane` and add a string property.
2. An `VisualObject` class. I implement it in [here](https://github.com/RileyGe/Macad3D/blob/9aac2f8dec861e7498ff84b4d79425c917d8e6ff/Source/Macad.Interaction/Visual/VisualText.cs), I think it is very simple too. I use `AIS_TextLable` as the `AisObject` and update code in mothed `void _UpdatePresentation()`. You can find the `void _UpdatePresentation()` below.
```csharp
void _UpdatePresentation()
{
    if (aisObject == null)
        return;
    aisObject.SetPosition(textLabel.Position);
    TCollection_ExtendedString text = new(textLabel.Text);
    aisObject.SetText(text);
    aisObject.SetFont("Arial");
    aisObject.SetHeight(20);
}
```
3. Registe the `VisualText` class to show the `TextLabel`. I implement this function in `VisualText` class, please find the code below.
```csharp
[AutoRegister]
internal static void Register()
{
    VisualObjectManager.Register<TextLabel>(Create);
}
```

I think these are all steps I need to do, but after I add the a `TextLabel` entity to the Document, but nothing happened. To do this, I add an `ActionCommand` in [AuxiliaryCommands](https://github.com/RileyGe/Macad3D/blob/main/Source/Macad.Interaction/Commands/AuxiliaryCommands.cs), you can find the code below:
```csharp
public static ActionCommand CreateTextLabel { get; } = new(
    () =>
    {
        TextLabel text = new()
        {
            Text = "Test Text",
            Position = new Pnt(0, 0, 0),
            //Rotation = _WorkspaceController.Workspace.WorkingPlane.Rotation()
        };                
        InteractiveContext.Current?.Document.Add(text);
        InteractiveContext.Current?.UndoHandler.Commit();

        //_WorkspaceController.Selection.SelectEntity(text);
        _WorkspaceController.Invalidate();
    },
    () => _CanStartTool())
{
    Header = () => "Text Label",
    Description = () => "Creates a new datum plane which can be used as a reference. A datum plan can be papered with an image.",
    Icon = () => "Auxiliary-DatumPlane",
    HelpTopic = "322f5cc2-0fc7-43f9-bb80-5e87cb3e3651"
};
```

I try to fix this problem for almost one day, give me some advise please, thank you very much.

**Reply by RileyGe** [reply]
Sorry for your time, I found the mistake myself. After I will update how to fix it.

**Comment by RileyGe** [comment]
The above three steps are basically correct but step 2 need to update.

1. When I implement the `VisualObject` class, we shoud set the Layer property.
2. Last answer I think `void _UpdatePresentation()` method will be call every time when we update the view port, but I was wrong. This method will be call only when the text show again after disappear.
3. We should Update/Display of the AISContext.

You can find the updated code [here](https://github.com/RileyGe/Macad3D/blob/7ccd9a7529a9bdff79aea579917f02a4512e9fb8/Source/Macad.Interaction/Visual/VisualText.cs)

Now I can show the text, but not action right when scaling and rotating. I will update the code lately.
